### PR TITLE
Update decompose hdf5

### PR DESCRIPTION
**Pull request by frheault**
This major update is simply to convert the output of `scil_decompose_connectivity.py`. Right now, this script output up to 25 000 files (for Brainnetome) and this means we will never be able to run large connectivity analysis on clusters due to file limits (Even worst if we want a Nextflow pipeline for connectivity).

In order to overcome this problem, the `.trk` output is now optional (for debugging and small investigation), but the main output is now a `.h5` that contains streamlines (IN VOX/CORNER SPACE) related to each edge.

`scil_compute_connectivity.py` has been adapted to support this new input. And a new script was added to compute a population average for each edge (input is a bunch of `.h5`) so the --similarity option works the same way as before.
This does not slow down the script, there is less chance of crashing due to the big reduction in IO.

PS : If you want data to test it I can provide some. But I think first @mdesco should try it and then we will see

### CODE REVIEW

File: scripts/scil_compute_connectivity.py [hunk 1]
```diff
@@ -108,46 +97,48 @@ def _processing_wrapper(args):
...
-            if not is_header_compatible(metric_filename, sft):
-                logging.error('{} and {} do not have a compatible header'.format(
-                    in_filename, metric_filename))
+            if not is_header_compatible(metric_filename, labels_img):
+                logging.error('{} do not have a compatible header'.format(
+                    metric_filename))
                 raise IOError
 
-            metric_data = nib.load(metric_filename).get_data()
+            metric_data = nib.load(metric_filename).get_fdata(dtype=np.float32)
```
Comment (arnaudbore) [inline scripts/scil_compute_connectivity.py]: np.float64 ?

File: scripts/scil_compute_connectivity.py [hunk 2]
```diff
@@ -164,11 +155,11 @@ def _build_arg_parser():
...
         description=__doc__,
         formatter_class=argparse.RawTextHelpFormatter,)
-    p.add_argument('in_bundles_dir',
-                   help='Folder containing all the bundle files (.trk).')
-    p.add_argument('labels_list',
-                   help='Text file containing the list of labels from the '
-                        'atlas.')
+    p.add_argument('in_hdf5',
+                   help='Input filename for the hdf5 container (.h5).')
+    p.add_argument('labels',
```
Comment (arnaudbore) [inline scripts/scil_compute_connectivity.py:160]: in_labels

File: scripts/scil_compute_connectivity.py [hunk 3]
```diff
@@ -280,12 +270,14 @@ def main():
...
+        total_elem = len(labels_list)**2 - len(labels_list)
+        results_elem = len(measures_dict.keys())*2 - len(labels_list)
     else:
-        total_elem = len(measures_dict.keys())*2 - len(labels_list)
+        total_elem = len(labels_list)**2
+        results_elem = len(measures_dict.keys())*2
 
-    logging.info('Out of {} node, {} contain values'.format(
-        total_elem, len(measures_dict.keys())*2))
+    logging.info('Out of {} possible nodes, {} contain values'.format(
```
Comment (arnaudbore) [inline scripts/scil_compute_connectivity.py:281]: values.

File: scripts/scil_decompose_connectivity.py [hunk 4]
```diff
@@ -127,11 +142,10 @@ def build_arg_parser():
                    help='Tractogram filename. Format must be one of \n'
                         'trk, tck, vtk, fib, dpy.')
     p.add_argument('labels',
-                   help='Labels file name (nifti).\nLabels must be consecutive '
-                        'from 0 to N, with 0 the background.\n'
+                   help='Labels file name (nifti).\n'
```
Comment (arnaudbore) [inline scripts/scil_decompose_connectivity.py]: in_labels

File: scripts/scil_compute_hdf5_average_density_map.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+        nib.save(nib.Nifti1Image(density_data, template_img.affine),
+                 os.path.join(out_dir, '{}.nii.gz'.format(key)))
+
+
+def main():
+    parser = _build_arg_parser()
+    args = parser.parse_args()
+
+    assert_inputs_exist(parser, args.in_hdf5+[args.population_template])
```
Comment (arnaudbore) [inline scripts/scil_compute_hdf5_average_density_map.py:83]: pep8 ?

File: scripts/scil_convert_tractogram.py [hunk 6]
```diff
@@ -43,8 +43,8 @@ def main():
     in_extension = os.path.splitext(args.in_tractogram)[1]
     out_extension = os.path.splitext(args.output_name)[1]
 
-    if in_extension == out_extension:
-        parser.error('Input and output cannot be of the same file format')
+    # if in_extension == out_extension:
+    #     parser.error('Input and output cannot be of the same file format')
```
Comment (arnaudbore) [inline scripts/scil_convert_tractogram.py:47]: ??

File: scripts/scil_decompose_connectivity.py [hunk 7]
```diff
@@ -168,6 +182,8 @@ def build_arg_parser():
                          'filtering with QB. [%(default)s]')
 
     s = p.add_argument_group('Saving options')
+    s.add_argument('--out_dir',
+                   help='Output directory for the streamlines representation.')
     s.add_argument('--save_raw_connections', action='store_true',
                    help='If set, will save all raw cut connections in a '
                         'subdirectory')
```
Comment (arnaudbore) [inline scripts/scil_decompose_connectivity.py:189]: `.` everywhere it's missing.

File: scripts/scil_compute_connectivity.py [hunk 8]
```diff
@@ -250,24 +238,27 @@ def main():
             # This is necessary to support more than one map for weighting
             measures_to_compute.append(in_name)
             dict_metrics_out_name[in_name] = out_name
-            measures_output_filename.append(out_name)
```
Comment (frheault) [inline scripts/scil_compute_connectivity.py:253]: To put back, sorry

File: scripts/scil_decompose_connectivity.py [hunk 9]
```diff
@@ -130,18 +144,18 @@ def build_arg_parser():
     p.add_argument('in_labels',
                    help='Labels file name (nifti). Labels must have 0 as '
                         'background.')
-    p.add_argument('output_dir',
+    p.add_argument('out_hdf5',
                    help='Output directory path.')
```
Comment (mdesco) [inline scripts/scil_decompose_connectivity.py:148]: Also, provide extension of the file please. I thought it was .hdf5 but when I see your next script (scil_compute_connectivity),  the input seem to want a .h5 file as input ...

File: scripts/scil_compute_connectivity.py [hunk 10]
```diff
@@ -163,11 +155,11 @@ def _build_arg_parser():
     p = argparse.ArgumentParser(
         description=__doc__,
         formatter_class=argparse.RawTextHelpFormatter,)
-    p.add_argument('in_bundles_dir',
-                   help='Folder containing all the bundle files (.trk).')
-    p.add_argument('labels_list',
-                   help='Text file containing the list of labels from the '
-                        'atlas.')
+    p.add_argument('in_hdf5',
+                   help='Input filename for the hdf5 container (.h5).')
```
Comment (mdesco) [inline scripts/scil_compute_connectivity.py:159]: Input filename for the hdf5 container coming from scil_decompose_connectivity (.h5) 

(if you want)

File: scripts/scil_compute_hdf5_average_density_map.py [hunk 11]
```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env python
+# encoding: utf-8
+
+"""
+Compute a density map for each edges of a connection from a HDF5 file.
+Typically use after scil_decompose_connectivity.py in order to obtain the
```
Comment (mdesco) [inline scripts/scil_compute_hdf5_average_density_map.py:6]: used

File: scripts/scil_compute_hdf5_average_density_map.py [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+average density map of each connections to enable the use of --similarity
+in scil_compute_connectivity.py.
+
+This script is parallelize, but will run much slower on non-SSD if too many
```
Comment (mdesco) [inline scripts/scil_compute_hdf5_average_density_map.py:10]: parallelized

File: scripts/scil_compute_hdf5_average_density_map.py [hunk 13]
```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env python
+# encoding: utf-8
+
+"""
+Compute a density map for each edges of a connection from a HDF5 file.
```
Comment (mdesco) [inline scripts/scil_compute_hdf5_average_density_map.py:5]: hdf5 in minusccule to be uniform and coherent with other scripts

File: scripts/scil_compute_hdf5_average_density_map.py [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+                             assert_inputs_exist,
+                             assert_output_dirs_exist_and_empty)
+from scilpy.tractanalysis.streamlines_metrics import compute_tract_counts_map
+
+
+def _build_arg_parser():
+    p = argparse.ArgumentParser(
+        description=__doc__, formatter_class=argparse.RawTextHelpFormatter)
+    p.add_argument('in_hdf5', nargs='+',
+                   help='List of HDF5 filenames (.h5) containing streamlines '
```
Comment (mdesco) [inline scripts/scil_compute_hdf5_average_density_map.py:36]: hdf5

File: scripts/scil_compute_hdf5_average_density_map.py [hunk 15]
```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env python
+# encoding: utf-8
+
+"""
+Compute a density map for each edges of a connection from a HDF5 file.
+Typically use after scil_decompose_connectivity.py in order to obtain the
+average density map of each connections to enable the use of --similarity
+in scil_compute_connectivity.py.
```
Comment (mdesco) [inline scripts/scil_compute_hdf5_average_density_map.py:8]: Honestly, after reading this small paragraph, I'm not sure what this script is about to do... Maybe lets discuss in the slack...

File: scripts/scil_compute_hdf5_average_density_map.py [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+                             assert_output_dirs_exist_and_empty)
+from scilpy.tractanalysis.streamlines_metrics import compute_tract_counts_map
+
+
+def _build_arg_parser():
+    p = argparse.ArgumentParser(
+        description=__doc__, formatter_class=argparse.RawTextHelpFormatter)
+    p.add_argument('in_hdf5', nargs='+',
+                   help='List of HDF5 filenames (.h5) containing streamlines '
+                        'data, offsets and lengths.')
```
Comment (mdesco) [inline scripts/scil_compute_hdf5_average_density_map.py:37]: coming from scil_decompose_connectivity

File: scripts/scil_compute_hdf5_average_density_map.py (hunk 15 above)
Comment (mdesco) [reply scripts/scil_compute_hdf5_average_density_map.py:8]: Just read input arguments needed and still not sure... 

Maybe give a simple example to illustrate what it does?

File: scripts/scil_compute_hdf5_average_density_map.py (hunk 13 above)
Comment (mdesco) [reply scripts/scil_compute_hdf5_average_density_map.py:5]: not sure what "for each edges of a connection" means

Comment (mdesco) [reply scripts/scil_compute_hdf5_average_density_map.py:5]: edges -> edge

File: scripts/scil_compute_hdf5_average_density_map.py (hunk 15 above)
Comment (mdesco) [reply scripts/scil_compute_hdf5_average_density_map.py:8]: connections -> connection

File: scripts/scil_compute_hdf5_average_density_map.py (hunk 11 above)
Comment (arnaudbore) [reply scripts/scil_compute_hdf5_average_density_map.py:6]: @frheault ?

File: scripts/scil_compute_connectivity.py [hunk 17]
```diff
@@ -38,67 +38,60 @@
...
-        if not is_header_compatible(in_filename, ref_filename):
-            logging.error('{} and {} do not have a compatible header'.format(
-                in_filename, ref_filename))
-            raise IOError
-        return nib.load(in_filename).get_fdata(dtype=np.float64)
+    if os.path.isfile(in_filename):
+        if not is_header_compatible(in_filename, ref_img):
+            raise IOError('{} do not have a compatible header'.format(
+                in_filename))
+        return nib.load(in_filename).get_fdata()
```
Comment (arnaudbore) [inline scripts/scil_compute_connectivity.py:63]: np.float64 or np.float32

**Comment by mdesco** [comment]
Ping. You are not waiting for me on right?
It is ready to go and quite important for our work with Sami.

Thanks!

**Pull request merged**